tests/abi/tools/revng-check-compatibility-with-abi/Verify.cpp

```diff
@@ -230,114 +227,299 @@ void VH::verify(const abi::FunctionType::Layout &FunctionLayout,
                             "expected.";
       revng_abort(Error.c_str());
     }
+  }
+}
+using VH = VerificationHelper;
+void VH::returnValue(const abi::FunctionType::Layout &FunctionLayout,
+                     const RVTest &Iterations) const {
+  auto PointerSize = model::Architecture::getPointerSize(Architecture);
+
+  for (const abi::runtime_test::ReturnValueTest &Test : *Iterations) {
+    const auto &Found = Test.ReturnValue;
+    const auto &Expected = Test.ExpectedReturnValue;
+    if (Found.Bytes.size() != Expected.Bytes.size()) {
+      std::string Error = "Return value size of '" + FunctionName.str()
+                          + "' is not consistent: something is *very* wrong "
+                            "with the test toolchain.";
+      revng_abort(Error.c_str());
+    }
 
-    // Check the return value.
-    auto RVR = FunctionLayout.returnValueRegisters();
-    auto ReturnValueRegisterList = ABI.sortReturnValues(std::move(RVR));
-    if (!Helper.Iteration.ReturnValue.Bytes.empty()) {
-      if (ReturnValueRegisterList.size() == 0) {
-        std::string Error = "Verification of the return value of '"
+    std::uint64_t MatchingByteCount = 0;
+    for (auto [FByte, EByte] : llvm::zip(Found.Bytes, Expected.Bytes))
+      if (FByte == EByte)
+        ++MatchingByteCount;
+
+    // Sadly, we cannot do a full comparison since return value structs often
+    // contain padding, value of which is not guaranteed to be preserved,
+    // causing the test to fail. We either need a way to tell if a specific byte
+    // belongs to the padding or not (which would restrict stuff we can test,
+    // for example, nested structs) or assume that non-perfect matches are
+    // acceptable.
+    //
+    // We know for sure that maximum padding amount is
+    // "maximum_possible_alignment_for_the_architecture - 1", which is achieved
+    // when a struct like this is returned:
+    // ```cpp
+    // struct ReturnValue {
+    //   field_with_maximum_alignment a; // let's say the alignment is 4.
+    //   field_with_minimum_alignment b; // the alignment always is 1.
+    //   // invisible 3 bytes here (4 - 1).
+    // };
+    // ```
+    // Which adds up to 3/8 (~37.5%). In the case where maximum alignment is
+    // 16 bytes - this jumps to 15/32 (~46.9%), and so on. Knowing that,
+    // we can assume that, at the very least, _more_ than half of the bytes
+    // will match.
+    //
+    // Also, considering we do multiple iterations of each test, the chance
+    // of having false positives is extremely low as well.
+    if (MatchingByteCount * 2 < Found.Bytes.size()) {
+      std::string Error = "Return value of '" + FunctionName.str()
+                          + "' was lost: something went *very* wrong.";
+      revng_abort(Error.c_str());
+    }
+
+    // Because we have no way to represent functions that use SPTAR in `rft`
```

Comment:
Wait, why are we testing RFTs at all?

Reply:
To make sure they are not broken: it adds very little complexity (since everything is based on a layout anyway) but gives me a lot of piece of mind.

On top of that, it also serves as the test that layout works well for RFTs too.

P. S. This comment might be slightly misleading, it doesn't only affects RFTs: it's exactly the same for CFTs after they have been converted down, I was just trying to stress that "we need this extra logic because CFT->RFT loses information".

Reply:
1. I do not understand the workflow for testing RFTs and I do not understand if we have any tests that actually does that.
Please, this is already complex enough, I really see no point in testing RFTs.
2. The purpose of this tool is to ensure that a `Layout` create from CFT correctly predicts where arguments/return values will be, correct?
3. If 2) is correct, what's the `CFG -> RFT` conversion matters at all?

Please remove all the complexity that's not strictly necessary for the goals of this tool.

Reply:
2. The original purpose of this tool is to make sure that no matter what weirdness happens inside either conversion, the prototype is still valid - as in it doesn't contradict the stuff we got by ~~ab~~using the compiler.

Since then it also shifted somewhat to also serve as the test for the `abi::FunctionType::Layout`  infrastructure too. Consider that we use Layouts all over the place, having them untested doesn't sit right with me.

But that's beside a point. Here's the only place where such information loss is even relevant. And it's fixed by a really simple workaround, which is what I was trying to say by the comment above (I guess using the term "RFT" there was a mistake). Here's a copy of literally all the "extra" complexity introduced for this:
```cpp
  bool UsesSPTAR = FunctionLayout.returnsAggregateType();
  if (!UsesSPTAR) {
    bool SingleArgument = FunctionLayout.Arguments.size() == 1;
    bool SingleReturnValue = FunctionLayout.ReturnValues.size() == 1;
    if (SingleArgument && SingleReturnValue) {
      const auto &ArgumentType = FunctionLayout.Arguments[0].Type;
      const auto &ReturnValueType = FunctionLayout.ReturnValues[0].Type;
      bool TypesMatch = ArgumentType == ReturnValueType;
      bool ReturnValueIsAPointer = ReturnValueType.isPointer();

      // Case for a register SPTAR - types must match.
      if (TypesMatch && ReturnValueIsAPointer)
        UsesSPTAR = true;

      // Case for a stack SPTAR - allow any type as the argument as long as
      // it's pointer-sized.
      if (*ArgumentType.size() == PointerSize && ReturnValueIsAPointer)
        UsesSPTAR = true;
    }
  }
```
It's a simple check that if the first argument is a pointer, and so is the return value, and that their types match.

Not to say, it's also perfectly safe, since tests currently only _either_ have arguments or return values. So having both - is a really good sign that it's a `SPTAR`-function.